### PR TITLE
<fix> Add private key null check in monitor.dart

### PR DESCRIPTION
**Pull request by JeremyTubongbanua**
**- What I did**
Added null check for `monitor.dart`. 

**- How I did it**
See line 255

**- How to verify it**
The following code now works. (Before it would throw a null). 
```dart
  AtClientPreference preference = AtClientPreference()
    ..cramSecret = cram
    ..commitLogPath = 'test/commitLogPath'
    ..hiveStoragePath = 'test/hive'
    ..isLocalStoreRequired = true;

  AtClientManager atClientManager = await AtClientManager.getInstance().setCurrentAtSign('fascinatingsnow', 'beaverarmadillo', preference);
  AtClient atClient = atClientManager.atClient;
  await atClient.getRemoteSecondary()!.authenticateCram(cram);
  List<String> keys = await atClient.getRemoteSecondary()!.atLookUp.scan();
  _logger.info(keys); // [@fascinatingsnow:signing_privatekey@fascinatingsnow, public:signing_publickey@fascinatingsnow]
```

**- Description for the changelog**
Authenticating CRAM was never possible through the `AtClientManager`. My change makes it possible.

### CODE REVIEW

**Review by gkc — APPROVED** [review]
Thanks, Jeremy! LGTM

**Comment by gkc** [comment]
Our cicd test images may be a bit borked, hence test failures. We are investigating.

**Comment by gkc** [comment]
@JeremyTubongbanua please merge the latest changes from atsign-foundation/at_client_sdk/trunk into your branch ... those changes should resolve the failing checks

**Comment by JeremyTubongbanua** [comment]
> @JeremyTubongbanua please merge the latest changes from atsign-foundation/at_client_sdk/trunk into your branch ... those changes should resolve the failing checks

I pulled the latest code from the at_client_sdk repo, and I still get this when trying to use `at_client`.
```
C:\Users\madiv\Desktop\GitHub\at_client>dart at_client/test/jeremy_test.dart
INFO|2022-05-20 15:25:49.379103|HiveBase|commit_log_b245a0bfe8a50a6f969ee3ac512173750bba572548b1c2a080aa6762766ca4da initialized successfully 

INFO|2022-05-20 15:25:49.398122|HiveBase|b245a0bfe8a50a6f969ee3ac512173750bba572548b1c2a080aa6762766ca4da initialized successfully 

Unhandled exception:
Null check operator used on a null value
#0      Monitor._authenticateConnection (package:at_client/src/manager/monitor.dart:255:62)
<asynchronous suspension>
#1      Monitor.start (package:at_client/src/manager/monitor.dart:176:7)
<asynchronous suspension>
#2      NotificationServiceImpl._startMonitor (package:at_client/src/service/notification_service_impl.dart:84:5)
<asynchronous suspension>
#3      NotificationServiceImpl._init (package:at_client/src/service/notification_service_impl.dart:53:5)
<asynchronous suspension>
#4      NotificationServiceImpl.create (package:at_client/src/service/notification_service_impl.dart:39:5)
<asynchronous suspension>
#5      AtClientManager.setCurrentAtSign (package:at_client/src/manager/at_client_manager.dart:61:27)
<asynchronous suspension>
#6      thing3 (file:///C:/Users/madiv/Desktop/GitHub/at_client/at_client/test/jeremy_test.dart:31:37)
<asynchronous suspension>
#7      main (file:///C:/Users/madiv/Desktop/GitHub/at_client/at_client/test/jeremy_test.dart:39:3)
<asynchronous suspension>
```

From running this code:
```dart
import 'package:at_client/at_client.dart';
import 'package:at_lookup/at_lookup.dart';

const cram = '123mycram123';

Future<void> thing3() async {
  AtClientPreference preference = AtClientPreference()
    ..cramSecret = cram
    ..commitLogPath = 'test/commitLogPath'
    ..hiveStoragePath = 'test/hive'
    ..isLocalStoreRequired = true;

  AtClientManager atClientManager = await AtClientManager.getInstance().setCurrentAtSign('fascinatingsnow', 'beaverarmadillo', preference);
  AtClient atClient = atClientManager.atClient;
  await atClient.getRemoteSecondary()!.authenticateCram(cram);
  List<String> keys = await atClient.getRemoteSecondary()!.atLookUp.scan();
  print(keys); // [@fascinatingsnow:signing_privatekey@fascinatingsnow, public:signing_publickey@fascinatingsnow]
}

void main() async {
  await thing3();
}
```

**Comment by JeremyTubongbanua** [comment]
> > @JeremyTubongbanua please merge the latest changes from atsign-foundation/at_client_sdk/trunk into your branch ... those changes should resolve the failing checks
> 
> I pulled the latest code from the at_client_sdk repo, and I still get this when trying to use `at_client`.
> 
> ```
> C:\Users\madiv\Desktop\GitHub\at_client>dart at_client/test/jeremy_test.dart
> INFO|2022-05-20 15:25:49.379103|HiveBase|commit_log_b245a0bfe8a50a6f969ee3ac512173750bba572548b1c2a080aa6762766ca4da initialized successfully 
> 
> INFO|2022-05-20 15:25:49.398122|HiveBase|b245a0bfe8a50a6f969ee3ac512173750bba572548b1c2a080aa6762766ca4da initialized successfully 
> 
> Unhandled exception:
> Null check operator used on a null value
> #0      Monitor._authenticateConnection (package:at_client/src/manager/monitor.dart:255:62)
> <asynchronous suspension>
> #1      Monitor.start (package:at_client/src/manager/monitor.dart:176:7)
> <asynchronous suspension>
> #2      NotificationServiceImpl._startMonitor (package:at_client/src/service/notification_service_impl.dart:84:5)
> <asynchronous suspension>
> #3      NotificationServiceImpl._init (package:at_client/src/service/notification_service_impl.dart:53:5)
> <asynchronous suspension>
> #4      NotificationServiceImpl.create (package:at_client/src/service/notification_service_impl.dart:39:5)
> <asynchronous suspension>
> #5      AtClientManager.setCurrentAtSign (package:at_client/src/manager/at_client_manager.dart:61:27)
> <asynchronous suspension>
> #6      thing3 (file:///C:/Users/madiv/Desktop/GitHub/at_client/at_client/test/jeremy_test.dart:31:37)
> <asynchronous suspension>
> #7      main (file:///C:/Users/madiv/Desktop/GitHub/at_client/at_client/test/jeremy_test.dart:39:3)
> <asynchronous suspension>
> ```
> 
> From running this code:
> 
> ```dart
> import 'package:at_client/at_client.dart';
> import 'package:at_lookup/at_lookup.dart';
> 
> const cram = '123mycram123';
> 
> Future<void> thing3() async {
>   AtClientPreference preference = AtClientPreference()
>     ..cramSecret = cram
>     ..commitLogPath = 'test/commitLogPath'
>     ..hiveStoragePath = 'test/hive'
>     ..isLocalStoreRequired = true;
> 
>   AtClientManager atClientManager = await AtClientManager.getInstance().setCurrentAtSign('fascinatingsnow', 'beaverarmadillo', preference);
>   AtClient atClient = atClientManager.atClient;
>   await atClient.getRemoteSecondary()!.authenticateCram(cram);
>   List<String> keys = await atClient.getRemoteSecondary()!.atLookUp.scan();
>   print(keys); // [@fascinatingsnow:signing_privatekey@fascinatingsnow, public:signing_publickey@fascinatingsnow]
> }
> 
> void main() async {
>   await thing3();
> }
> ```

Oops. My apologies @gkc. You said **merge**. Let me get on that.

**Comment by JeremyTubongbanua** [comment]
@gkc There. I think I've done it. (I've never done this before! haha)

**Comment by gkc** [comment]
Yeah because I don't have permission to merge the latest changes from the main trunk into your branch

**Comment by gkc** [comment]
Thanks Jeremy. Sorry about this; something else is wrong also. We're in the middle of moving to a new set of atsigns for testing, entirely dedicated to this repo and well, it's tricky 😀🤷‍♂️

I'll ping you again once I'm sure that another merge from trunk is certain to work.

**Comment by JeremyTubongbanua** [comment]
> Thanks Jeremy. Sorry about this; something else is wrong also. We're in the middle of moving to a new set of atsigns for testing, entirely dedicated to this repo and well, it's tricky 😀🤷‍♂️
> 
> I'll ping you again once I'm sure that another merge from trunk is certain to work.

Take your time @gkc. :) Good luck!

**Comment by gkc** [comment]
@JeremyTubongbanua  OK ... https://github.com/atsign-foundation/at_client_sdk/pull/512 has been merged to trunk. Please merge from this trunk into your cram_fix branch again - thanks!

**Comment by JeremyTubongbanua** [comment]
> @JeremyTubongbanua OK ... #512 has been merged to trunk. Please merge from this trunk into your cram_fix branch again - thanks!

How can I do that on my cmd terminal?

File: at_client/lib/src/manager/monitor.dart [hunk 1]
```diff
@@ -252,6 +252,7 @@ class Monitor {
     }
     _logger.finer(
         'Authenticating the monitor connection: from result:$fromResponse');
+    if(_preference.privateKey == null) throw Exception('Private key is null in preference'); // JEREMY
```
Comment (gkc) [inline at_client/lib/src/manager/monitor.dart]: @JeremyTubongbanua please remove `// JEREMY` ... git will tell us who did the commit 😀

**Review by gkc — CHANGES_REQUESTED** [review]

**Comment by gkc** [comment]
Hi @JeremyTubongbanua,

The e2e tests are failing for your PR because of access control on github secrets. We need to configure the actions so we only run the 'unit tests' check (which also actually includes the functional tests, we need to split it up into two checks) for pull requests from external branches.

File: at_client/lib/src/manager/monitor.dart (hunk 1 above)
Comment (JeremyTubongbanua) [reply at_client/lib/src/manager/monitor.dart]: See f070f6a8de74c0368e89b9f238615a1b8751a364 👍

**Comment by JeremyTubongbanua** [comment]
> Hi @JeremyTubongbanua,
> 
> The e2e tests are failing for your PR because of access control on github secrets. We need to configure the actions so we only run the 'unit tests' check (which also actually includes the functional tests, we need to split it up into two checks) for pull requests from external branches.

Okay
So you want me to make a `cram_fix_unit_test` branch and a `cram_fix_e2e_tests` branch and make pull requests for both of them?

**Comment by gkc** [comment]
No, we'll take care of it Jeremy - thanks!

**Comment by gkc** [comment]
I believe this PR was obsoleted by another one from the same repo

**Pull request closed without merging**